### PR TITLE
Adding support for running integration tests from golang

### PR DESCRIPTION
**Pull request by nmittler**
Adding support for running integration tests from golang
    
Generally useful for helping to debug integration tests. This gives us
the option of running the integration tests within an IDE
(e.g. GoLand).
    
All of the pilot integration test code has been moved to
tests/e2e/tests/pilot. Both `main.go` and `driver.go` have been
removed in favor of the new `pilot_test.go`. I've also updated
make to use the new tests, rather than the old integration test
binary.

### CODE REVIEW

File: pilot/test/integration/driver_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	params.TestDataDir = "testdata/"
+	params.CertDir = "../../docker/certs/"
+
+	// Configure the docker HUB to use for images.
+	// TODO(nmittler): Use local HUB and build images locally.
+	params.Hub = "gcr.io/istio-testing"
+	params.Tag = "54a3ecc8ae3e72c919d63436d0d12a7c5f3600c8"
+
+	// Run all the tests.
+	if err := runTests(*params); err != nil {
```
Comment (costinm) [inline pilot/test/integration/driver_test.go]: Can you pass t as parameter, and have the runTest method do t.Run() ? 
If I read this correctly, it'll show up as a single big golang test - which is not much better than the original. Fine to do this in a separate PR, of course - but if something is planned please add TODOs so reviewers know...

File: pilot/test/integration/main.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	flag.BoolVar(&params.debugImagesAndMode, "debug", params.debugImagesAndMode,
+		"Use debug images and mode (false for prod)")
+	flag.BoolVar(&params.SkipCleanup, "skip-cleanup", params.SkipCleanup,
+		"Debug, skip clean up")
+	flag.BoolVar(&params.SkipCleanupOnFailure, "skip-cleanup-on-failure", params.SkipCleanupOnFailure,
+		"Debug, skip clean up on failure")
+}
+
+// Main entry point for integration test executable.
+func main() {
```
Comment (costinm) [inline pilot/test/integration/main.go]: Confused, if you do 'go test ..' - how do the flags get set ? 
Also, why do we still have the main - is it for transition ? Again, please add TODOs or indicate the intentions...

For new tests I would expect people to just write normal go test, and in time existing test would be converted - I don't mind a transition period, but having both main and test may confuse people.

File: pilot/test/integration/test_api.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+// Test is the interface for all integration tests.
+type test interface {
```
Comment (costinm) [inline pilot/test/integration/test_api.go]: Same here - please add deprecated, todo, etc.

File: pilot/test/integration/main.go (hunk 2 above)
Comment (nmittler) [reply pilot/test/integration/main.go]: The point of this PR is to maintain what we have today, which is an executable that runs our integration tests.  The `main.go` simply moves the main function to its own file, this isn't new. 

What's new in this PR is `driver_test.go`, which is a golang test.  If you were to run `go test` (and had the appropriate environment variables set) this is what would be run (not `main`).

I expect that we'll eventually remove `main.go` in favor of `driver_test.go`, but we need to make sure that all uses of `main` are swapped over to use golang tests first.

**Comment by costinm** [comment]
It's very unlikely the tests will work in circleci as is - they are not unit tests, but the get picked by the makefile.  To get pass this, I see 3 options:
- move the files to top level tests/integration (next to the other integration tests - in future we want to converge, including reusing some of the common methods to setup and fetch logs fro the e2e tests)
- exclude the tests with some regex or make trick
- add a "UNIT_TEST" env variable, and in the test skip if the env is set.

Probably other options exist, I would prefer 1 or 3.

**Comment by costinm** [comment]
I don't mind keeping main in the short term - but it should be clearly
marked as deprecated (maybe print a message), and it should
have some comments indicating on what is missing from 'go test' and what is
the plan.

In particular I don't see how the flags in main are reflected in the 'go
test' mode - so I doubt anyone could really use the go test
in practical ways ( and unit tests are already broken in circleci ). I
haven't check prow pre-submit time, AFAIK it's also running
all tests so now will try to run pilot integration (and twice)

File: pilot/test/integration/driver_test.go (hunk 1 above)
Comment (nmittler) [reply pilot/test/integration/driver_test.go]: Absolutely ... good idea.  I've restructured things so that this test is no longer dependent on driver.go ... instead, it restructures the `runTests` method to use `testing.T`.

File: pilot/test/integration/test_api.go (hunk 3 above)
Comment (nmittler) [reply pilot/test/integration/test_api.go]: Added a TODO.

**Comment by nmittler** [comment]
@costinm PTAL

I've restructured the integration test to use `testing.T` for the individual test classes.  I've also added an environment variable `RUN_ISTIO_INTEG_TESTS` to control whether the integration tests are run.

**Comment by kyessenov** [comment]
I have two suggestions that can be done later:
1. Test binaries allow coverage collection. That would be a wonderful way to bump up coverage if we can compile pilot into a test binary.
2. Can we use some guard _not to run_ environment-depending targets as part of regular `go test`. I think that go generally trusts that tests are isolated (e.g. do not share environment like a cluster) and self-contained (since test results are cached, a change in the cluster should invalidate the cache).

**Comment by ayj** [comment]
> That would be a wonderful way to bump up coverage if we can compile pilot into a test binary.

https://groups.google.com/a/google.com/d/msg/istio-mixer/UzrQAFyFvHE/jeKvTJpXDAAJ could be used to run a local instance of pilot built as a test binary. Additional firewall rules may be needed for this approach.

> Can we use some guard not to run environment-depending targets as part of regular go test. 

Build tags are one option here, e.g. 

```golang
// +build test_extern_env
```
```bash                 
go test -tags test_extern_env
```

**Comment by nmittler** [comment]
@ayj, @kyessenov PTAL

@ayj good suggestion on using build tags ... didn't know they existed :). I've added an `integration` build tag to the file and converted the test to a `TestMain` ... not sure it really matters, but it seemed more appropriate. I'm also skipping the tests when `testing.Short()`.

**Comment by nmittler** [comment]
> https://groups.google.com/a/google.com/d/msg/istio-mixer/UzrQAFyFvHE/jeKvTJpXDAAJ could be used to run a local instance of pilot built as a test binary. Additional firewall rules may be needed for this approach.

It's absolutely on my radar to support running Pilot/Mixer locally as this framework evolves.  I've already been talking with Zack about it ... stay tuned.

**Comment by kyessenov** [comment]
Could you document build tags in developer guide if you do decide to use them?

File: pilot/test/integration/infra.go [hunk 4]
```diff
@@ -110,6 +133,49 @@ const (
...
+		TestCount:             1,
+		SelectedTest:          "",
+		debugImagesAndMode:    true,
+		UseAutomaticInjection: false,
+		UseAdmissionWebhook:   false,
+		AdmissionServiceName:  "istio-pilot",
+	}
+}
+
+func (infra *infra) copyWithDefaultAuth() *infra {
```
Comment (ayj) [inline pilot/test/integration/infra.go]: The function name implies something is copied, but the code is mutating the `infra` object and returning it. Maybe something like the following?

```golang
out := *infra
out.Name = ....
return &out
```

File: pilot/test/integration/infra.go [hunk 5]
```diff
@@ -643,25 +715,46 @@ func (infra *infra) createSidecarInjector() error {
...
+	if filledYaml, err := infra.fill("sidecar-injector.yaml.tmpl", infra); err != nil {
 		log.Infof("Sidecar injector template could not be processed, please delete stale injector webhook: %v",
 			err)
-	} else if err = infra.kubeDelete(yaml, infra.IstioNamespace); err != nil {
+	} else if err = infra.kubeDelete(filledYaml, infra.IstioNamespace); err != nil {
 		log.Infof("Sidecar injector could not be deleted: %v", err)
 	}
 }
+
+// fill a file based on a template
```
Comment (ayj) [inline pilot/test/integration/infra.go]: As an aside, I wonder if test code would be more maintainable if we created the k8s resources directly with client-go instead of some combination of `text/template` and shelling out to kubectl. For example, see the configmap creation for mutating webhooks below. It's immediately clear what test values are used during resource creation vs. applying a template over the `infra` object which selectively uses some (but not all) of the fields.

https://github.com/istio/istio/blob/89e56e8fe7c08923a10c3a3dcdba90924a3eb1b8/pilot/test/integration/infra.go#L617-L626

Comment (nmittler) [reply pilot/test/integration/infra.go]: Definitely seems like something worth exploring ... care to file a bug and assign to me?

**Comment by costinm** [comment]
Sorry, didn't see this - IMHO build tags create more problems for the build
system ( we need to build the test 2x), and little benefit since
we really just need to exclude the entire directory. They are meant for
other things - platform specific code, optional features.

Just move the entire directory to istio/test - next to all other
integration tests. This may also allow more code reuse and visibility.

I think it is critical to get rid of the current intialization in pilot
tests, so the setup step can be shared and be consistent across
all  e2e tests ( e2e_simple in particular ), it will also fit with some of
the automation that k8s is doing on their side.

**Comment by costinm** [comment]
To clarify: we spent a lot of time to optimize the CI pipeline, I don't
mind removing the tag in a separate PR, but long
term moving the code is a better solution. IMHO it would be cleaner/better
to just keep all integration tests together,
but if you want to do it in separate PR it's ok.

**Comment by nmittler** [comment]
@costinm I totally agree.  My plan was to evolve this into a more flexible framework that could be used for all of our larger tests, so I think we're on the same page.  Perhaps we can chat offline?

WRT tags, I don't think you need to build and test 2x ... if you're building on a system that is set up for k8s, you simply build once with the `integration` tag.  All tests would be built and run in that case (including the integration tests). Regardless, for now, having the tag properly excludes it from the default build, which is what we want.  We can discuss the best way to enable/disable integration tests once we replace the current `main` entirely.  Sound reasonable?

**Comment by rshriram** [comment]
can we invoke pilot e2e tests in circleci using make test .. instead of using pilot/test/integration/.. ? Heck, we could even remove building those test binaries. But will that cause Go to rebuild the entire code in e2e-pilot ? That would be a bummer, as we optimized the build pipeline using workspaces for this very purpose.

**Comment by nmittler** [comment]
@costinm @rshriram @ayj I've reworked this a bit to move all of the files to `tests/e2e/tests/pilot`, as suggested by @costinm.  This allowed me to remove the use of tags and also makes the tests more consistent with the others (albeit without use of the common framework ... baby steps :)).

**Comment by ldemailly** [comment]
linters not happy:
```
tests/e2e/tests/pilot/auth_exclusion.go:23:1:error: authExclusion is unused (deadcode)
tests/e2e/tests/pilot/egress_rules.go:30:1:error: egressRules is unused (deadcode)
tests/e2e/tests/pilot/grpc.go:23:1:error: grpc is unused (deadcode)
tests/e2e/tests/pilot/headless.go:23:1:error: headless is unused (deadcode)
tests/e2e/tests/pilot/http.go:25:1:error: http is unused (deadcode)
tests/e2e/tests/pilot/infra.go:140:1:error: newInfra is unused (deadcode)
tests/e2e/tests/pilot/ingress.go:28:1:error: ingress is unused (deadcode)
tests/e2e/tests/pilot/kubernetes_external_name_services.go:23:1:error: kubernetesExternalNameServices is unused (deadcode)
tests/e2e/tests/pilot/routing.go:31:1:error: routing is unused (deadcode)
tests/e2e/tests/pilot/routingToEgress.go:32:1:error: routingToEgress is unused (deadcode)
tests/e2e/tests/pilot/tcp.go:24:1:error: tcp is unused (deadcode)
tests/e2e/tests/pilot/test_api.go:19:1:error: test is unused (deadcode)
tests/e2e/tests/pilot/zipkin.go:30:1:error: zipkin is unused (deadcode)
Makefile:241: recipe for target 'lint' failed
make: *** [lint] Error 1
```

**Comment by nmittler** [comment]
@costinm mind re-approving now that the route tests have been fixed?

File: tests/istio.mk [hunk 6]
```diff
@@ -86,6 +86,5 @@ e2e_upgrade: istioctl generate_yaml
 
 e2e_all: e2e_simple e2e_mixer e2e_bookinfo
 
-e2e_pilot:
-	go install ./pilot/test/integration
-	${ISTIO_BIN}/integration --logtostderr ${TESTOPTS} -hub ${HUB} -tag ${TAG}
+e2e_pilot: istioctl generate_yaml
+	go test -v -timeout 20m ./tests/e2e/tests/pilot -hub ${HUB} -tag ${TAG}
```
Comment (mattdelco) [inline tests/istio.mk]: Both prow (in prow/istio-pilot-e2e.sh) and circleCI (in .circleci/config.yml) run e2e_pilot with a bunch of parameters passed in via TESTOPTS (which is now removed) so more work might be needed here (or elsewhere).

Comment (nmittler) [reply tests/istio.mk]: Good catch @mattdelco ... created #3459 to fix

Comment (andraxylia) [reply tests/istio.mk]: So how do I run this now with parameters?

Comment (nmittler) [reply tests/istio.mk]: @andraxylia you just add the parameters to the "program arguments" in GoLang (or whatever IDE you're using) ... or from the command line with `go test <dir> <args>`
